From 8ae65cd9d0bc4edd29cd31cc801d8c7d5162db18 Mon Sep 17 00:00:00 2001
From: Gareth Kitchen <gkitchen@cookability.biz>
Date: Tue, 17 Jul 2018 09:58:29 +0100
Subject: [PATCH 2/4] Localise NTP to improve GPS TTFF for non-North American
 users.

Default NTP Server is set here:
https://github.com/LineageOS/android_frameworks_base/blob/lineage-15.1/core
/res/res/values/config.xml#L1847 Currently: time.android.com

Testing response times to this server from around the globe reveals in ms:-

Europe			 <30
Middle East		 <68
North America		<150
Johannesburg		 183
Buenos Aires		 220
Tokyo			 226
Sydney			 276
Hong Kong		 285
Brisbane		 295
Mumbai			 349
Beijing			4691
Shanghai		4906
Russia			 n/a

Poor GPS TTFF are frequently mentioned in support channels.

It's clear that the default Android timeserver cannot deliver and for some
reason the framework hardcodes the GPS specific NTP server to default to:
north-america.pool.ntp.org

Great for North America and it appears Europe but it does not address the
global issue. Also, the pool.ntp.org project forbids both hardware and
software vendors from using these default zone names.
http://www.pool.ntp.org/en/vendors.html

To address the Asia problems it look like a special country based config
file for mcc460 (China) has been created to overlay asia.pool.ntp.org.

So it makes sense to leverage the ntp.org's existing 'android' vendor name
to make the default ntp server for GPS purposes:
1.android.pool.ntp.org

This will return a random but accurate NTP server in close geopraphic
proximity to the device.

Testing on my own build in the UK seems to improve hot and cold TTFF
considerably.

Change-Id: I144af45757efa35b32daf034eece6e046d2bde79
---
 gps/etc/gps.conf                                   | 14 +++++---------
 .../base/core/res/res/values-mcc460/config.xml     |  1 -
 .../frameworks/base/core/res/res/values/config.xml | 12 +++++++++++-
 3 files changed, 16 insertions(+), 11 deletions(-)
 mode change 100644 => 100755 gps/etc/gps.conf
 mode change 100644 => 100755 overlay/frameworks/base/core/res/res/values-mcc460/config.xml
 mode change 100644 => 100755 overlay/frameworks/base/core/res/res/values/config.xml

diff --git a/gps/etc/gps.conf b/gps/etc/gps.conf
old mode 100644
new mode 100755
index 8834810..eae4ba8
--- a/gps/etc/gps.conf
+++ b/gps/etc/gps.conf
@@ -14,19 +14,15 @@ XTRA_VERSION_CHECK=1
 # _CLEAR = 0
 ERR_ESTIMATE=0
 
-#Test
-# NTP_SERVER=time.izatcloud.net
-#Asia
-# NTP_SERVER=asia.pool.ntp.org
-#Europe
-# NTP_SERVER=europe.pool.ntp.org
-#North America
-#NTP_SERVER=north-america.pool.ntp.org
+# The http://pool.ntp.org project is a big virtual cluster of 
+# timeservers. The device will receive the IP address of an NTP 
+# Server in close proximity to its location.
+#NTP_SERVER=1.android.pool.ntp.org
 
 # DEBUG LEVELS: 0 - none, 1 - Error, 2 - Warning, 3 - Info
 #               4 - Debug, 5 - Verbose
 # If DEBUG_LEVEL is commented, Android's logging levels will be used
-DEBUG_LEVEL = 2
+DEBUG_LEVEL = 4
 
 # Intermediate position report, 1=enable, 0=disable
 INTERMEDIATE_POS=0
diff --git a/overlay/frameworks/base/core/res/res/values-mcc460/config.xml b/overlay/frameworks/base/core/res/res/values-mcc460/config.xml
old mode 100644
new mode 100755
index 3a2cbc6..46223bd
--- a/overlay/frameworks/base/core/res/res/values-mcc460/config.xml
+++ b/overlay/frameworks/base/core/res/res/values-mcc460/config.xml
@@ -26,7 +26,6 @@
         <item>XTRA_SERVER_1=https://xtrapath1.izatcloud.net/xtra3grc.bin</item>
         <item>XTRA_SERVER_2=https://xtrapath2.izatcloud.net/xtra3grc.bin</item>
         <item>XTRA_SERVER_3=https://xtrapath3.izatcloud.net/xtra3grc.bin</item>
-        <item>NTP_SERVER=asia.pool.ntp.org</item>
         <item>SUPL_ES=0</item>
         <item>SUPL_HOST=supl.nokia.com</item>
         <item>SUPL_PORT=7275</item>
diff --git a/overlay/frameworks/base/core/res/res/values/config.xml b/overlay/frameworks/base/core/res/res/values/config.xml
old mode 100644
new mode 100755
index bc00ebc..b7aa1bb
--- a/overlay/frameworks/base/core/res/res/values/config.xml
+++ b/overlay/frameworks/base/core/res/res/values/config.xml
@@ -19,6 +19,16 @@
 
 <!-- These resources are around just to allow their values to be customized
      for different hardware and product builds. -->
+
+<!-- The http://pool.ntp.org project is a big virtual cluster of timeservers.
+	 The device will receive the IP address of an NTP Server in close proximity
+	 to its location. 
+	 
+	 Operating System and Software vendors must absolutely not use the default
+	 pool.ntp.org zone names. So it makes sense to leverage the pool's existing
+	 Android vendor name.
+	 -->
+
 <resources>
 
     <!-- Values for GPS configuration -->
@@ -26,7 +36,7 @@
         <item>XTRA_SERVER_1=https://xtrapath1.izatcloud.net/xtra3grc.bin</item>
         <item>XTRA_SERVER_2=https://xtrapath2.izatcloud.net/xtra3grc.bin</item>
         <item>XTRA_SERVER_3=https://xtrapath3.izatcloud.net/xtra3grc.bin</item>
-        <item>NTP_SERVER=north-america.pool.ntp.org</item>
+        <item>NTP_SERVER=1.android.pool.ntp.org</item>
         <item>SUPL_ES=0</item>
         <item>SUPL_HOST=supl.google.com</item>
         <item>SUPL_PORT=7275</item>
-- 
2.17.1

